src/Kinto.elm

```diff
@@ -0,0 +1,316 @@
+
+withHeader : String -> String -> Config -> Config
+withHeader name value ({ headers } as config) =
+    let
+        newHeaders =
+            ( name, value ) :: headers
+    in
+        { config | headers = newHeaders }
+
+
+withAuthHeader : Auth -> Config -> Result Error Config
+withAuthHeader auth config =
+    case auth of
+        NoAuth ->
+            Ok config
+
+        Basic username password ->
+            let
+                hash str =
+                    Base64.encode str |> Result.formatError ConfigError
+
+                addAuthHeader hash =
+                    Ok (withHeader "Authorization" ("Basic " ++ hash) config)
+            in
+                hash (username ++ ":" ++ password)
+                    `Result.andThen` addAuthHeader
+
+        Bearer token ->
+            Ok (config |> withHeader "Authorization" ("Bearer " ++ token))
+
+
+configure : Url -> Auth -> Result Error Config
+configure baseUrl auth =
+    Config baseUrl []
+        |> withAuthHeader auth
+        |> Debug.log "Kinto request"
+
+
+
+-- Dealing with answers from the Kinto server
+
+
+bodyDataDecoder : Json.Decoder Json.Value
+bodyDataDecoder =
+    ("data" := Json.value)
+
+
+extractData : String -> StatusCode -> StatusMsg -> Task Error Json.Value
+extractData body statusCode statusMsg =
+    let
+        responseResult =
+            Json.decodeString bodyDataDecoder body
+    in
+        case responseResult of
+            Ok data ->
+                Task.succeed data
+
+            Err _ ->
+                -- Is it an ErrorRecord json?
+                decodeError body statusCode statusMsg
+
+
+errorDecoder : Json.Decoder ErrorRecord
+errorDecoder =
+    Json.object4 ErrorRecord
+        ("errno" := Json.int)
+        ("message" := Json.string)
+        ("code" := Json.int)
+        ("error" := Json.string)
+
+
+decodeError : String -> StatusCode -> StatusMsg -> Task Error a
+decodeError body statusCode statusMsg =
+    case Json.decodeString errorDecoder body of
+        Ok errorRecord ->
+            Task.fail (KintoError statusCode statusMsg errorRecord)
+
+        Err msg ->
+            Task.fail (ServerError statusCode statusMsg body)
+
+
+toKintoResponse : Task Http.RawError Http.Response -> Task Error Json.Value
+toKintoResponse response =
+    (Task.mapError NetworkError response)
+        `Task.andThen` (handleResponse extractData)
+
+
+handleResponse :
+    (String -> StatusCode -> StatusMsg -> Task Error Json.Value)
+    -> Http.Response
+    -> Task Error Json.Value
+handleResponse handle response =
+    case response.value of
+        Http.Text body ->
+            handle body response.status response.statusText
+
+        _ ->
+            Task.fail
+                (ServerError
+                    response.status
+                    response.statusText
+                    "Response body is a blob"
+                )
+
+
+
+-- High level API
+
+
+performQuery : Result Error Config -> Endpoint -> Verb -> Task Error Json.Value
```

Comment:
For now we're dealing with invalid config for this strange base64 encoding case, but soon we'll have to validate the server url as well (eg. it should be http(s), have the version, etc., as we do in other client implementations). So providing invalid configuration values should definitely fail imho, so the user knows about what's wrong :)

May we just want to `type alias Config = Result Error ValidConfig` or something?